src/effects/backends/builtin/glitcheffect.cpp

```diff
@@ -0,0 +1,124 @@
+    quantize->setId("quantize");
+    quantize->setName(QObject::tr("Quantize"));
+    quantize->setShortName(QObject::tr("Quantize"));
+    quantize->setDescription(QObject::tr(
+            "Round the Time parameter to the nearest 1/4 beat."));
+    quantize->setValueScaler(EffectManifestParameter::ValueScaler::Toggle);
+    quantize->setUnitsHint(EffectManifestParameter::UnitsHint::Unknown);
+    quantize->setRange(0, 1, 1);
+
+    EffectManifestParameterPointer triplet = pManifest->addParameter();
+    triplet->setId("triplet");
+    triplet->setName(QObject::tr("Triplets"));
+    triplet->setShortName(QObject::tr("Triplets"));
+    triplet->setDescription(
+            QObject::tr("When the Quantize parameter is enabled, divide "
+                        "rounded 1/4 beats of Time parameter by 3."));
+    triplet->setValueScaler(EffectManifestParameter::ValueScaler::Toggle);
+    triplet->setUnitsHint(EffectManifestParameter::UnitsHint::Unknown);
+    triplet->setRange(0, 0, 1);
+
+    return pManifest;
+}
+
+void GlitchEffect::loadEngineEffectParameters(
+        const QMap<QString, EngineEffectParameterPointer>& parameters) {
+    m_pDelayParameter = parameters.value("delay_time");
+    m_pQuantizeParameter = parameters.value("quantize");
+    m_pTripletParameter = parameters.value("triplet");
+}
+
+void GlitchEffect::processChannel(
+        GlitchGroupState* pGroupState,
+        const CSAMPLE* pInput,
+        CSAMPLE* pOutput,
+        const mixxx::EngineParameters& engineParameters,
+        const EffectEnableState enableState,
+        const GroupFeatureState& groupFeatures) {
+
+    // The minimum of the parameter is zero so the exact center of the knob is 1 beat.
+    double period = m_pDelayParameter->value();
+
+    int delay_frames;
+    double max_delay;
+    double min_delay;
+    if (groupFeatures.has_beat_length_sec) {
+        // period is a number of beats
+        if (m_pQuantizeParameter->toBool()) {
+            period = std::max(roundToFraction(period, 4), 1 / 8.0);
+            if (m_pTripletParameter->toBool()) {
+                period /= 3.0;
+            }
+        } else if (period < 1 / 8.0) {
+            period = 1 / 8.0;
+        }
+        delay_frames = static_cast<int>(period * groupFeatures.beat_length_sec *
+                engineParameters.sampleRate());
+        max_delay = 2 * groupFeatures.beat_length_sec * engineParameters.sampleRate();
+        min_delay = 1 / 8.0 * groupFeatures.beat_length_sec * engineParameters.sampleRate();
+    } else {
+        // period is a number of seconds
+        period = std::max(period, 1 / 8.0);
+        delay_frames = static_cast<int>(period * engineParameters.sampleRate());
+        max_delay = 2 * engineParameters.sampleRate();
+        min_delay = 1 / 8.0 * engineParameters.sampleRate();
+    }
+
+    // Scale the delay so that the knob starts at no glitch and ends at the maximum delay.
+    delay_frames = static_cast<int>((delay_frames - min_delay) *
```

Comment:
I'm not sure this is quite right. Doesn't this effectively change the range of delay_frames from `(min_delay, max_delay)` to `(0, max_delay-min_delay)`?

I think what you want is the following (along with my other code suggestions):

```suggestion
    if (delay_frames < min_delay) {
        delay_frames = 0;
    }
```

Reply:
Furthermore, you can bypass the effect here by instead doing:

```c++
if (delay_frames < min_delay) {
    SampleUtil::copy(
            pOutput,
            pInput,
            engineParameters.samplesPerBuffer());
    return;
}
```

This avoids one buffer copy (from pInput to repeat_buf).